src/contracts/core/RewardsCoordinator.sol

```diff
@@ -229,35 +238,65 @@ contract RewardsCoordinator is
         onlyWhenNotPaused(PAUSED_REWARD_OPERATOR_SET)
         nonReentrant
     {
-        for (uint256 i = 0; i < rewardsSubmissions.length;) {
+        // Cache starting nonce value for later use.
+        uint256 startingNonce = submissionNonce[msg.sender];
+        for (uint256 i = 0; i < rewardsSubmissions.length; ++i) {
             OperatorSetRewardsSubmission calldata rewardsSubmission = rewardsSubmissions[i];
-            uint256 nonce = submissionNonce[msg.sender];
-            bytes32 rewardsSubmissionHash = keccak256(abi.encode(msg.sender, nonce, rewardsSubmission));
-
+            // Calculate current nonce by adding `i` to starting nonce.
+            uint256 currentNonce = startingNonce + i;
+            // Calculate rewards submission hash.
+            bytes32 rewardsSubmissionHash = keccak256(abi.encode(msg.sender, currentNonce, rewardsSubmission));
+            // Assert operator set is valid/registered for the given avs (msg.sender).
             require(
                 avsDirectory.isOperatorSet(msg.sender, rewardsSubmission.operatorSetId),
                 "RewardsCoordinator.rewardOperatorSetForRange: invalid operatorSet"
             );
+            // Validate rewards submission.
             _validateRewardsSubmission(
                 rewardsSubmission.strategiesAndMultipliers,
-                rewardsSubmission.token,
                 rewardsSubmission.amount,
                 rewardsSubmission.startTimestamp,
                 rewardsSubmission.duration,
                 OPERATOR_SET_MAX_RETROACTIVE_LENGTH,
                 OPERATOR_SET_GENESIS_REWARDS_TIMESTAMP
             );
-
+            // Mark `rewardsSubmissionHash` as valid for avs (msg.sender).
             isAVSRewardsSubmissionHash[msg.sender][rewardsSubmissionHash] = true;
-            submissionNonce[msg.sender] = nonce + 1;
-
-            emit OperatorSetRewardCreated(msg.sender, nonce, rewardsSubmissionHash, rewardsSubmission);
+            // Pull reward `token` of `amount` from avs (msg.sender).
             rewardsSubmission.token.safeTransferFrom(msg.sender, address(this), rewardsSubmission.amount);
-
-            unchecked {
-                ++i;
-            }
+            // Emit event to track reward submission.
+            emit OperatorSetRewardCreated(msg.sender, currentNonce, rewardsSubmissionHash, rewardsSubmission);
         }
+        // Increase avs submission nonce by rewards submissions submitted.
+        submissionNonce[msg.sender] = startingNonce + rewardsSubmissions.length;
+    }
+
+    function rewardOperatorsForPerformance(PerformanceRewardsSubmission calldata rewardsSubmission)
+        external
+        onlyWhenNotPaused(PAUSED_REWARD_PERFORMANCE)
+        nonReentrant
+    {
+        // Cache nonce for later use.
```

Comment:
Nit: the `startTimestamp + duration` should be < block.timestamp as we want the reward to be completely retroactive.

Reply:
Sounds like we need another check in _validateRewardsSubmission()?

Reply:
Nah, because only this reward cannot go in the future. The rest of the rewards can. Doesn't make sense to give a reward for performance in the future if there is no performance yet :)